src/main.clj

```diff
@@ -1,2 +1,314 @@
+; decurry reduces a curry of functions to a single function
+; The single function takes variable number of args
+(((fn [f]
+    (fn [& a]
+      (reduce #(% %2) f a)))
+  (fn [a]
+    (fn [b]
+      (fn [c]
+        (fn [d]
+          (* a b c d))))))
+ 1 2 3 4)
+
+; https://4clojure.oxal.org/#/problem/177
+; Couldn't solve myself. Had to look up solutions.
+; Understood that string when passed to reduce is broken into characters.
+#(= ()
+    (reduce
+      (fn [a x]
+        (if a (case x
+                (\[ \{ \() (conj a x)
+                (\] \} \))
+                (if (= ({\] \[ \} \{ \) \(} x) (first a))
+                  (rest a) nil)
+                a)))
+      () %))
+
+; Kitty's solution
+(defn balanced? [s]
+  (let [pairs {\( \) \{ \} \[ \]} ;; Map is chosen here because it's versatile.
+        openers (set (keys pairs)) ;; Using sets here,
+        closers (set (vals pairs)) ;; And here, so we can use contains? later.
+        ;; Create a set by combining the openers and closers.
+        search-set (into openers closers)
+        ;; We can now simply filter the string against the set,
+        ;; To return just the pair-occurrences that we care about.
+        occurrences (filter #(contains? search-set %) s)]
+    (empty? ;; If the height-stack is empty, stuff should be balanced.
+      (reduce (fn [height-stack occurrence]
+                (cond
+                  ;; (contains?) reads better than get or shorthand
+                  (contains? openers occurrence)
+                  (conj height-stack occurrence)
+                  ;; Checks if the closing occurrence closes the top of the height-collecting stack.
+                  ;; If so, pop the top.
+                  ;; (peek) is more efficient than (last) on a vector.
+                  (and (contains? closers occurrence)
+                       (= (get pairs (peek height-stack)) occurrence))
+                  (pop height-stack)
+                  ;; This is sort of a null-case, for dangling closers.
+                  ;; It's efficient to exit-early here, hence wrap in (reduced)
+                  :else (reduced (conj height-stack occurrence))))
+              []
+              occurrences))))
+; I wanted to solve this by regex pattern matching,
+; Realized that iterating over is much more efficient.
+
+; https://4clojure.oxal.org/#/problem/76
+; Writing trampoline is easy
+; Check if return value is funciton using function?
+; Do recursion accordingly.
+(= [1 3 5 7 9 11]
+   (letfn
+     [(foo [x y] #(bar (conj x y) y))
+      (bar [x y] (if (> (last x) 10)
+                   x
+                   #(foo x (+ 2 y))))]
+     (trampoline foo [] 1)))
+
+; https://4clojure.oxal.org/#/problem/106
+; Double, Halve or Add 2
+; Unsolved
+
+; https://4clojure.oxal.org/#/problem/73
+; This solution isn't extensible :( 😭
+((fn tic-tac-toe
+   ([board] (tic-tac-toe 0 board))
+   ([c board]
+    ; Iterate over all rows to find a winner.
+    (let [winner (reduce (fn [_ r]
```

Comment:
I think this `reduce` can definitely be changed to a `filter`. It will short-circuit.